File: server/src/crt.ts
import { spawn } from "node:child_process";
import * as crypto from "node:crypto";
import * as fs from "node:fs";

function temp_name(): string {
    return `/tmp/${crypto.randomBytes(48).toString("hex")}`;
}

export function strip(crt: string): string {
    let ans = "";
    for (const line of crt.split("\n")) {
        if (!line.startsWith("-----") || !line.endsWith("-----")) ans += line;
    }
    return ans;
}

export function generate_key(): Promise<string> {
    return new Promise<string>((res, rej) => {
        const p = spawn(
            "openssl",
            ["ecparam", "-name", "prime256v1", "-genkey", "-noout", "-out", "-"],
            { stdio: ["ignore", "pipe", "inherit"] },
        );
        if (p.stdout === null) throw Error("should not be null");
        let key = "";
        p.stdout.on("data", (data) => {
            key += data;
        });
        p.on("close", (code) => {
            if (code === 0 && key) res(key);
            else rej("Failed");
        });
    });
}

export function generate_ca_crt(key: string): Promise<string> {
    return new Promise<string>((res, rej) => {
        const t1 = temp_name();
        fs.writeFileSync(
            t1,
            "[req]\nprompt = no\ndistinguished_name = distinguished_name\n[distinguished_name]\nC=US\n",
            { mode: 0o600 },
        );
        const p = spawn(
            "openssl",
            [
                "req",
                "-x509",
                "-new",
                "-nodes",
                "-key",
                "-",
                "-sha256",
                "-days",
                "9999",
                "-out",
                "-",
                "-config",
                t1,
            ],
            { stdio: ["pipe", "pipe", "inherit"] },
        );
        if (p.stdin === null) throw Error("should not be null");
        p.stdin.write(key, () => p.stdin?.end());
        if (p.stdout === null) throw Error("should not be null");
        let crt = "";
        p.stdout.on("data", (data) => {
            crt += data;
        });
        p.on("close", (code) => {
            fs.unlink(t1, () => {});
            if (code === 0 && crt) res(crt);
            else rej("Failed");
        });
    });
}

export function generate_srs(key: string, cn: string): Promise<string> {
    return new Promise<string>((res, rej) => {
        const t1 = temp_name();
        fs.writeFileSync(
            t1,
            `[req]\nprompt = no\ndistinguished_name = distinguished_name\n[distinguished_name]\nCN=${cn}\n`,
            { mode: 0o400 },
        );
        const t2 = temp_name();
        fs.writeFileSync(t2, key, { mode: 0o400 });
        const p = spawn("openssl", ["req", "-new", "-key", t2, "-out", "-", "-config", t1], {
            stdio: ["pipe", "pipe", "inherit"],
        });
        if (p.stdout === null) throw Error("should not be null");
        let srs = "";
        p.stdout.on("data", (data) => {
            srs += data;
        });
        p.on("close", (code) => {
            fs.unlink(t1, () => {});
            if (code === 0 && srs) res(srs);
            else rej("Failed");
        });
    });
}

export function generate_crt(
    ca_key: string,
    ca_crt: string,
    srs: string,
    subcerts: string[] = [],
    timeoutDays = 999,
): Promise<string> {
    return new Promise<string>((res, rej) => {
        const t1 = temp_name();
        const t2 = temp_name();
        const t4 = temp_name();
        fs.writeFileSync(t1, srs, { mode: 0o400 });
        fs.writeFileSync(t2, ca_crt, { mode: 0o400 });
        fs.writeFileSync(t4, ca_key, { mode: 0o400 });

        const args = [
            "x509",
            "-req",
            "-days",
            `${timeoutDays}`,
            "-in",
            t1,
            "-CA",
            t2,
            "-CAkey",
            t4,
            "-CAcreateserial",
            "-out",
            "-",
        ];
        let t3: string | null = null;
        if (subcerts.length) {
            t3 = temp_name();
            fs.writeFileSync(
                t3,
                `basicConstraints = critical, CA:TRUE\nkeyUsage = critical, keyCertSign, cRLSign, digitalSignature, nonRepudiation, keyEncipherment, keyAgreement\nsubjectKeyIdentifier = hash\nnameConstraints = critical, ${subcerts.map((v) => `permitted;DNS:${v}`).join(", ")}\n`,
                { mode: 0o400 },
            );
            args.push("-extfile");
            args.push(t3);
        }
        const p = spawn("openssl", args, { stdio: ["pipe", "pipe", "inherit"] });

        if (p.stdout === null) throw Error("should not be null");
        let crt = "";
        p.stdout.on("data", (data) => {
            crt += data;
        });
        p.on("close", (code) => {
            fs.unlink(t1, () => {});
            fs.unlink(t2, () => {});
            if (t3) {
                fs.unlink(t3, () => {});
            }
            if (code === 0 && crt) res(crt);
            else rej("Failed");
        });
    });
}

export async function generate_ssh_crt(
    keyId: string,
    principal: string,
    caPrivateKey: string,
    clientPublicKey: string,
    validityDays: number,
    type: "host" | "user",
): Promise<string> {
    const sshHostCaKeyFile = temp_name();
    const tmp = temp_name();
    const clientPublicKeyFile = `${tmp}.pub`;
    const outputCertificateFile = `${tmp}-cert.pub`;
    try {
        fs.writeFileSync(
            sshHostCaKeyFile,
            `-----BEGIN OPENSSH PRIVATE KEY-----\n${caPrivateKey}\n-----END OPENSSH PRIVATE KEY-----\n`,
            { mode: 0o400 },
        );
        fs.writeFileSync(clientPublicKeyFile, clientPublicKey, { mode: 0o400 });
        const args = [
            "-s",
            sshHostCaKeyFile,
            "-I",
            keyId,
            ...(type === "host" ? ["-h"] : []),
            "-n",
            principal,
            "-V",
            `-5m:+${+validityDays}d`,
            "-z",
            "42",
            clientPublicKeyFile,
        ];
        const p = spawn("ssh-keygen", args, { stdio: ["inherit", "inherit", "inherit"] });
        const code: number = await new Promise((r) => p.on("close", r));
        if (code !== 0) throw new Error(`ssh-keygen exited with code ${code}`);
        return fs.readFileSync(outputCertificateFile, { encoding: "utf-8" });
    } finally {
        fs.unlink(sshHostCaKeyFile, () => {});
        fs.unlink(clientPublicKeyFile, () => {});
        fs.unlink(outputCertificateFile, () => {});
    }
}

File: src/bin/server/crt.rs
use anyhow::{Result, bail};
use std::io::Write;
use tempfile::{NamedTempFile, TempDir};

pub fn strip(crt: &str) -> &str {
    let mut ans = crt;
    while let Some((head, rem)) = ans.split_once('\n') {
        if head.starts_with("-----") && head.ends_with("-----") {
            ans = rem;
        } else {
            break;
        }
    }
    while let Some((rem, tail)) = ans.rsplit_once("\n") {
        if tail.is_empty() || (tail.starts_with("-----") && tail.ends_with("-----")) {
            ans = rem;
        } else {
            break;
        }
    }
    ans
}

pub async fn generate_key() -> Result<String> {
    let res = tokio::process::Command::new("openssl")
        .args([
            "ecparam",
            "-name",
            "prime256v1",
            "-genkey",
            "-noout",
            "-out",
            "-",
        ])
        .stdin(std::process::Stdio::null())
        .stderr(std::process::Stdio::inherit())
        .stdout(std::process::Stdio::piped())
        .output()
        .await?;
    if !res.status.success() {
        bail!("openssl exited with code {} in generate_key", res.status);
    }
    Ok(String::from_utf8(res.stdout)?)
}

pub async fn generate_ca_crt(key: &str) -> Result<String> {
    let mut t1 = NamedTempFile::new()?;
    t1.write_all(
        "[req]\nprompt = no\ndistinguished_name = distinguished_name\n[distinguished_name]\nC=US\n"
            .as_bytes(),
    )?;

    let mut t2 = NamedTempFile::new()?;
    t2.write_all(key.as_bytes())?;

    let res = tokio::process::Command::new("openssl")
        .args(["req", "-x509", "-new", "-nodes", "-key"])
        .arg(t2.path())
        .args(["-sha256", "-days", "9999", "-out", "-", "-config"])
        .arg(t1.path())
        .stdin(std::process::Stdio::null())
        .stderr(std::process::Stdio::inherit())
        .stdout(std::process::Stdio::piped())
        .output()
        .await?;
    if !res.status.success() {
        bail!("openssl exited with code {} in generate_ca_crt", res.status);
    }
    Ok(String::from_utf8(res.stdout)?)
}

pub async fn generate_srs(key: &str, cn: &str) -> Result<String> {
    let mut t1 = NamedTempFile::new()?;
    write!(
        &mut t1,
        "[req]\nprompt = no\ndistinguished_name = distinguished_name\n[distinguished_name]\nCN={}\n",
        cn
    )?;
    let mut t2 = NamedTempFile::new()?;
    t2.write_all(key.as_bytes())?;
    let res = tokio::process::Command::new("openssl")
        .args(["req", "-new", "-key"])
        .arg(t2.path())
        .args(["-out", "-", "-config"])
        .arg(t1.path())
        .stdin(std::process::Stdio::null())
        .stderr(std::process::Stdio::inherit())
        .stdout(std::process::Stdio::piped())
        .output()
        .await?;
    if !res.status.success() {
        bail!("openssl exited with code {} in generate_srs", res.status);
    }
    Ok(String::from_utf8(res.stdout)?)
}

pub async fn generate_crt(
    ca_key: &str,
    ca_crt: &str,
    srs: &str,
    subcerts: &[String],
    timeout_days: u32,
) -> Result<String> {
    let mut t1 = NamedTempFile::new()?;
    let mut t2 = NamedTempFile::new()?;
    let mut t3 = NamedTempFile::new()?;
    let mut t4 = NamedTempFile::new()?;
    t1.write_all(srs.as_bytes())?;
    t2.write_all(ca_crt.as_bytes())?;
    t4.write_all(ca_key.as_bytes())?;

    let mut cmd = tokio::process::Command::new("openssl");
    cmd.arg("x509");
    cmd.arg("-req");
    cmd.arg("-days");
    cmd.arg(timeout_days.to_string());
    cmd.arg("-in");
    cmd.arg(t1.path());
    cmd.arg("-CA");
    cmd.arg(t2.path());
    cmd.arg("-CAkey");
    cmd.arg(t4.path());
    cmd.arg("-CAcreateserial");
    cmd.arg("-out");
    cmd.arg("-");
    if !subcerts.is_empty() {
        write!(&mut t3, "basicConstraints = critical, CA:TRUE
keyUsage = critical, keyCertSign, cRLSign, digitalSignature, nonRepudiation, keyEncipherment, keyAgreement
subjectKeyIdentifier = hash
nameConstraints = critical")?;
        for v in subcerts {
            write!(&mut t3, ", permitted;DNS:{}", v)?;
        }
        cmd.arg("-extfile");
        cmd.arg(t3.path());
    }
    let res = cmd
        .stdin(std::process::Stdio::null())
        .stderr(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .output()
        .await?;
    if !res.status.success() {
        bail!(
            "openssl exited with code {} in generate_crt:{}{}",
            res.status,
            String::from_utf8_lossy(&res.stdout),
            String::from_utf8_lossy(&res.stderr)
        );
    }
    Ok(String::from_utf8(res.stdout)?)
}

pub enum Type {
    Host,
    User,
}

pub async fn generate_ssh_crt(
    key_id: &str,
    principal: &str,
    ca_private_key: &str,
    client_public_key: &str,
    validity_days: u32,
    r#type: Type,
) -> Result<String> {
    let mut ssh_host_ca_key_file = NamedTempFile::with_suffix(".pem")?;
    let tmp_dir = TempDir::new()?;
    let client_public_key_path = tmp_dir.path().join("thing.pub");
    std::fs::write(&client_public_key_path, client_public_key.as_bytes())?;
    let output_certificate_path = tmp_dir.path().join("thing-cert.pub");
    writeln!(
        ssh_host_ca_key_file,
        "-----BEGIN OPENSSH PRIVATE KEY-----\n{}\n-----END OPENSSH PRIVATE KEY-----",
        ca_private_key
    )?;
    ssh_host_ca_key_file.flush()?;

    let mut cmd = tokio::process::Command::new("ssh-keygen");
    cmd.arg("-s");
    cmd.arg(ssh_host_ca_key_file.path());
    cmd.arg("-I");
    cmd.arg(key_id);
    if matches!(r#type, Type::Host) {
        cmd.arg("-h");
    }
    cmd.arg("-n");
    cmd.arg(principal);
    cmd.arg("-V");
    cmd.arg(format!("-5m:+{}d", validity_days));
    cmd.arg("-z");
    cmd.arg("42");
    cmd.arg(&client_public_key_path);
    let res = cmd
        .stdin(std::process::Stdio::null())
        .stderr(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .output()
        .await?;

    if !res.status.success() {
        bail!(
            "ssh-keygen exited with code {}: {}{}",
            res.status,
            String::from_utf8_lossy(&res.stdout),
            String::from_utf8_lossy(&res.stderr)
        );
    }
    Ok(std::fs::read_to_string(output_certificate_path)?)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_strip() {
        assert_eq!(
            strip("-----BEGIN EC PRIVATE KEY-----\nX\nY\n-----END EC PRIVATE KEY-----\n"),
            "X\nY"
        );
        assert_eq!(
            strip("-----BEGIN EC PRIVATE KEY-----\nX\nY\n-----END EC PRIVATE KEY-----"),
            "X\nY"
        );
    }
}
